cmd/jobsink/main.go

```diff
@@ -0,0 +1,353 @@
+
+	tlsConfig, err := getServerTLSConfig(ctx)
+	if err != nil {
+		log.Fatal("Failed to get TLS config", err)
+	}
+
+	sm, err := eventingtls.NewServerManager(ctx,
+		kncloudevents.NewHTTPEventReceiver(8080),
+		kncloudevents.NewHTTPEventReceiver(8443, kncloudevents.WithTLSConfig(tlsConfig)),
+		h,
+		configMapWatcher,
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := controller.StartInformers(ctx.Done(), informers...); err != nil {
+		logger.Fatal("Failed to start informers", zap.Error(err))
+	}
+
+	// Start informers and wait for them to sync.
+	logger.Info("Starting informers.")
+	if err := controller.StartInformers(ctx.Done(), informers...); err != nil {
+		logger.Fatal("Failed to start informers", zap.Error(err))
+	}
+
+	// Start the servers
+	logger.Info("Starting...")
+	if err = sm.StartServers(ctx); err != nil {
+		logger.Fatal("StartServers() returned an error", zap.Error(err))
+	}
+	tracer.Shutdown(context.Background())
+	logger.Info("Exiting...")
+}
+
+type Handler struct {
+	k8s    kubernetes.Interface
+	lister sinkslister.JobSinkLister
+	logger *zap.Logger
+}
+
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodGet {
+		h.handleGet(w, r)
+		return
+	}
+
+	if r.Method != http.MethodPost {
+		h.logger.Info("Unexpected HTTP method", zap.String("method", r.Method))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	parts := strings.Split(strings.TrimSuffix(r.RequestURI, "/"), "/")
+	if len(parts) != 3 {
+		h.logger.Info("Malformed uri", zap.String("URI", r.RequestURI), zap.Any("parts", parts))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	ref := types.NamespacedName{
+		Namespace: parts[1],
+		Name:      parts[2],
+	}
+
+	h.logger.Debug("Handling POST request", zap.String("URI", r.RequestURI))
+
+	message := cehttp.NewMessageFromHttpRequest(r)
+	defer message.Finish(nil)
+
+	event, err := binding.ToEvent(r.Context(), message)
+	if err != nil {
+		h.logger.Warn("failed to extract event from request", zap.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if err := event.Validate(); err != nil {
+		h.logger.Info("failed to validate event from request", zap.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	js, err := h.lister.JobSinks(ref.Namespace).Get(ref.Name)
+	if err != nil {
+		h.logger.Warn("Failed to retrieve jobsink", zap.String("ref", ref.String()), zap.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	jobs, err := h.k8s.BatchV1().Jobs(js.GetNamespace()).List(r.Context(), metav1.ListOptions{
+		LabelSelector: jobLabelSelector(ref, event.Source(), event.ID()),
+		Limit:         1,
+	})
+	if err != nil {
+		h.logger.Warn("Failed to retrieve job", zap.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if len(jobs.Items) > 0 {
+		w.Header().Add("Location", locationHeader(ref, event.Source(), event.ID()))
+		w.WriteHeader(http.StatusAccepted)
+		return
+	}
```

Comment:
What's the reasoning for returning here?

Reply:
This is handling idempotency when there is a job already scheduled for the event (uniquely identified with source + id), we return early